File: docker/backtesting-server/optimization/havwap-optuna-optimizer.js
/**
 * HAVWAP Strategy Optuna Hyperparameter Optimization
 * 
 * Optimizes HAVWAP-Proper strategy parameters using Optuna framework
 * Targets: Daily Sharpe Ratio maximization and Total PnL
 * Test Period: Multiple weeks of market data for robust parameter identification
 * Holding: All-day positions allowed for maximum opportunity capture
 */

const optuna = require('optuna');
const moment = require('moment-timezone');
const { BacktestEngine } = require('../engine/backtest-engine');
const HAVWAPProperStrategy = require('../strategies/havwap-proper');
const AlpacaClient = require('../utils/alpaca-client');

class HAVWAPOptunaOptimizer {
  constructor(config = {}) {
    this.config = {
      // Optimization settings
      nTrials: config.nTrials || 500, // Number of optimization trials
      nJobs: config.nJobs || 4, // Parallel jobs
      studyName: config.studyName || 'havwap-optimization',
      
      // Test period (multiple weeks for robust results)
      startDate: config.startDate || '2024-10-01',
      endDate: config.endDate || '2024-10-25',
      
      // Symbols to test (liquid 0DTE options)
      symbols: config.symbols || ['SPY', 'QQQ', 'IWM'],
      
      // Backtest settings
      initialCapital: config.initialCapital || 10000,
      maxPositions: config.maxPositions || 1,
      contracts: config.contracts || 1,
      
      // Optimization targets
      primaryMetric: config.primaryMetric || 'sharpe_ratio', // 'sharpe_ratio', 'total_return', 'profit_factor'
      secondaryMetric: config.secondaryMetric || 'total_return',
      
      // Risk constraints
      maxDrawdown: config.maxDrawdown || 0.20, // 20% max drawdown
      minTrades: config.minTrades || 10, // Minimum trades for valid optimization
      
      ...config
    };
    
    this.alpacaClient = new AlpacaClient();
    this.study = null;
    this.results = [];
  }

  /**
   * Define parameter search spaces for Optuna
   */
  defineParameterSpaces(trial) {
    return {
      // Distance thresholds (most critical parameters)
      d_entry: trial.suggest_float('d_entry', 0.0005, 0.0050, step=0.0001), // 0.05% - 0.50%
      d_near: trial.suggest_float('d_near', 0.0002, 0.0015, step=0.0001),   // 0.02% - 0.15%
      s_min: trial.suggest_float('s_min', 0.00001, 0.0005, step=0.00001),   // Slope sensitivity
      
      // Volatility filters
      rv_cap: trial.suggest_float('rv_cap', 0.002, 0.008, step=0.0005),     // 20-80 bps realized vol cap
      iv_chg_cap: trial.suggest_float('iv_chg_cap', 1.5, 5.0, step=0.25),   // IV change cap
      
      // Order flow threshold
      theta_imb: trial.suggest_float('theta_imb', 0.55, 0.80, step=0.05),   // Imbalance threshold
      
      // Delta targets
      deltaTargetReversion: trial.suggest_float('deltaTargetReversion', 0.40, 0.60, step=0.02),
      deltaTargetTrend: trial.suggest_float('deltaTargetTrend', 0.50, 0.65, step=0.02),
      
      // Exit parameters - REVERSION strategy
      profitTargetReversion: trial.suggest_float('profitTargetReversion', 0.10, 0.40, step=0.02), // 10-40%
      stopLossReversion: trial.suggest_float('stopLossReversion', 0.05, 0.20, step=0.01),         // 5-20%
      timeStopReversion: trial.suggest_int('timeStopReversion', 60, 600, step=30),                 // 1-10 minutes
      
      // Exit parameters - TREND strategy  
      profitTargetTrend: trial.suggest_float('profitTargetTrend', 0.12, 0.45, step=0.02),         // 12-45%
      stopLossTrend: trial.suggest_float('stopLossTrend', 0.08, 0.25, step=0.01),                 // 8-25%
      timeStopTrend: trial.suggest_int('timeStopTrend', 120, 900, step=30),                       // 2-15 minutes
      
      // Option selection
      maxSpreadPct: trial.suggest_int('maxSpreadPct', 8, 25, step=2),                             // Max bid/ask spread %
      minVolume: trial.suggest_int('minVolume', 0, 5, step=1),                                    // Minimum volume filter
      
      // All-day holding option
      zeroDTECloseTime: trial.suggest_categorical('zeroDTECloseTime', ['15:30', '15:45', '15:50', '15:55']) // Close timing
    };
  }

  /**
   * Get weekdays between start and end dates
   */
  getWeekdays(startDate, endDate) {
    const weekdays = [];
    const current = moment(startDate);
    const end = moment(endDate);
    
    while (current.isSameOrBefore(end)) {
      // Only include weekdays (Monday=1, Friday=5)
      if (current.isoWeekday() >= 1 && current.isoWeekday() <= 5) {
        weekdays.push(current.format('YYYY-MM-DD'));
      }
      current.add(1, 'day');
    }
    
    return weekdays;
  }

  /**
   * Calculate performance metrics from backtest results
   */
  calculateMetrics(results) {
    if (!results || results.length === 0) {
      return {
        total_return: 0,
        sharpe_ratio: 0,
        max_drawdown: 1,
        profit_factor: 0,
        win_rate: 0,
        total_trades: 0,
        avg_return_per_trade: 0,
        validity_score: 0
      };
    }

    const returns = [];
    const dailyPnL = [];
    let totalPnL = 0;
    let totalTrades = 0;
    let winningTrades = 0;
    let losingTrades = 0;
    let grossWins = 0;
    let grossLosses = 0;
    let runningBalance = this.config.initialCapital;
    let peak = this.config.initialCapital;
    let maxDrawdown = 0;

    // Process each day's results
    results.forEach(dayResult => {
      if (dayResult.trades && dayResult.trades.length > 0) {
        let dayPnL = 0;
        
        dayResult.trades.forEach(trade => {
          const tradePnL = trade.realized_pnl || 0;
          totalPnL += tradePnL;
          dayPnL += tradePnL;
          totalTrades++;
          
          if (tradePnL > 0) {
            winningTrades++;
            grossWins += tradePnL;
          } else {
            losingTrades++;
            grossLosses += Math.abs(tradePnL);
          }
        });
        
        dailyPnL.push(dayPnL);
        runningBalance += dayPnL;
        
        // Track drawdown
        if (runningBalance > peak) {
          peak = runningBalance;
        } else {
          const drawdown = (peak - runningBalance) / peak;
          maxDrawdown = Math.max(maxDrawdown, drawdown);
        }
        
        // Calculate daily return
        const dailyReturn = dayPnL / this.config.initialCapital;
        returns.push(dailyReturn);
      } else {
        // No trades day - zero return
        returns.push(0);
        dailyPnL.push(0);
      }
    });

    // Calculate metrics
    const totalReturn = totalPnL / this.config.initialCapital;
    const avgReturn = returns.reduce((a, b) => a + b, 0) / returns.length;
    const variance = returns.reduce((sum, ret) => sum + Math.pow(ret - avgReturn, 2), 0) / returns.length;
    const volatility = Math.sqrt(variance * 252); // Annualized
    const sharpeRatio = volatility > 0 ? (avgReturn * 252) / volatility : 0;
    
    const winRate = totalTrades > 0 ? winningTrades / totalTrades : 0;
    const profitFactor = grossLosses > 0 ? grossWins / grossLosses : grossWins > 0 ? 999 : 0;
    const avgReturnPerTrade = totalTrades > 0 ? totalPnL / totalTrades : 0;
    
    // Validity score (0-1) based on having sufficient trades and reasonable metrics
    let validityScore = 0;
    if (totalTrades >= this.config.minTrades) validityScore += 0.4;
    if (maxDrawdown <= this.config.maxDrawdown) validityScore += 0.3;
    if (sharpeRatio > 0) validityScore += 0.2;
    if (profitFactor > 1.0) validityScore += 0.1;

    return {
      total_return: totalReturn,
      sharpe_ratio: sharpeRatio,
      max_drawdown: maxDrawdown,
      profit_factor: profitFactor,
      win_rate: winRate,
      total_trades: totalTrades,
      avg_return_per_trade: avgReturnPerTrade,
      validity_score: validityScore,
      daily_pnl: dailyPnL,
      gross_wins: grossWins,
      gross_losses: grossLosses
    };
  }

  /**
   * Objective function for Optuna optimization
   */
  async objective(trial) {
    try {
      console.log(`\n🔬 OPTUNA Trial ${trial.number}:`);
      
      // Get parameter suggestions from Optuna
      const params = this.defineParameterSpaces(trial);
      console.log(`   Parameters: d_entry=${params.d_entry.toFixed(4)}, d_near=${params.d_near.toFixed(4)}, s_min=${params.s_min.toFixed(6)}`);
      console.log(`   Exits: Reversion PT=${params.profitTargetReversion.toFixed(2)}, SL=${params.stopLossReversion.toFixed(2)}, Time=${params.timeStopReversion}s`);
      
      // Get all weekdays in test period
      const weekdays = this.getWeekdays(this.config.startDate, this.config.endDate);
      console.log(`   Testing across ${weekdays.length} trading days`);
      
      const allResults = [];
      let totalBacktests = 0;
      let successfulBacktests = 0;
      
      // Test each symbol across all days
      for (const symbol of this.config.symbols) {
        for (const date of weekdays) {
          try {
            totalBacktests++;
            
            // Create strategy with trial parameters
            const strategy = new HAVWAPProperStrategy(params);
            
            // Create backtest engine
            const engine = new BacktestEngine(this.alpacaClient);
            
            // Run backtest for this day/symbol
            const result = await engine.runBacktest({
              strategy: strategy,
              symbol: symbol,
              startDate: date,
              endDate: date,
              initialCapital: this.config.initialCapital,
              maxPositions: this.config.maxPositions,
              contracts: this.config.contracts,
              mode: 'backtest'
            });
            
            if (result && result.success) {
              allResults.push(result);
              successfulBacktests++;
            }
            
          } catch (error) {
            console.log(`   ⚠️  Backtest failed for ${symbol} ${date}: ${error.message}`);
            // Continue with other tests
          }
        }
      }
      
      console.log(`   ✅ Completed ${successfulBacktests}/${totalBacktests} backtests`);
      
      // Calculate comprehensive metrics
      const metrics = this.calculateMetrics(allResults);
      
      console.log(`   📊 Results: Return=${(metrics.total_return*100).toFixed(2)}%, Sharpe=${metrics.sharpe_ratio.toFixed(3)}, Trades=${metrics.total_trades}`);
      console.log(`   📉 Risk: MaxDD=${(metrics.max_drawdown*100).toFixed(2)}%, PF=${metrics.profit_factor.toFixed(2)}, WinRate=${(metrics.win_rate*100).toFixed(1)}%`);
      
      // Store result for analysis
      this.results.push({
        trial_number: trial.number,
        parameters: params,
        metrics: metrics,
        successful_backtests: successfulBacktests,
        total_backtests: totalBacktests
      });
      
      // Optuna objective: Composite score
      let objectiveScore = 0;
      
      if (metrics.validity_score < 0.5) {
        // Penalize invalid parameter sets
        objectiveScore = -1000;
      } else {
        // Multi-objective optimization
        if (this.config.primaryMetric === 'sharpe_ratio') {
          objectiveScore = metrics.sharpe_ratio * 0.7 + metrics.total_return * 0.3;
        } else if (this.config.primaryMetric === 'total_return') {
          objectiveScore = metrics.total_return * 0.7 + metrics.sharpe_ratio * 0.3;
        } else if (this.config.primaryMetric === 'profit_factor') {
          objectiveScore = Math.log(metrics.profit_factor) * 0.6 + metrics.sharpe_ratio * 0.4;
        }
        
        // Penalty for excessive drawdown
        if (metrics.max_drawdown > this.config.maxDrawdown) {
          objectiveScore *= (1 - metrics.max_drawdown);
        }
      }
      
      console.log(`   🎯 Objective Score: ${objectiveScore.toFixed(4)}`);
      
      return objectiveScore;
      
    } catch (error) {
      console.error(`   ❌ Trial ${trial.number} failed: ${error.message}`);
      return -1000; // Return large penalty for failed trials
    }
  }

  /**
   * Run Optuna optimization study
   */
  async runOptimization() {
    console.log(`\n🚀 Starting HAVWAP Optuna Optimization:`);
    console.log(`   Study: ${this.config.studyName}`);
    console.log(`   Trials: ${this.config.nTrials}`);
    console.log(`   Period: ${this.config.startDate} to ${this.config.endDate}`);
    console.log(`   Symbols: ${this.config.symbols.join(', ')}`);
    console.log(`   Target: ${this.config.primaryMetric}`);
    
    try {
      // Create Optuna study
      this.study = optuna.create_study({
        study_name: this.config.studyName,
        direction: 'maximize', // Maximize objective score
        sampler: optuna.samplers.TPESampler({
          n_startup_trials: 50, // Random exploration first
          n_ei_candidates: 24,
          seed: 42
        }),
        pruner: optuna.pruners.MedianPruner({
          n_startup_trials: 25,
          n_warmup_steps: 5
        })
      });
      
      // Run optimization
      await this.study.optimize(this.objective.bind(this), {
        n_trials: this.config.nTrials,
        n_jobs: this.config.nJobs,
        timeout: 3600 * 24 // 24 hour timeout
      });
      
      console.log(`\n✅ Optimization completed!`);
      
      // Get best parameters
      const bestParams = this.study.best_params;
      const bestValue = this.study.best_value;
      
      console.log(`\n🏆 BEST PARAMETERS (Score: ${bestValue.toFixed(4)}):`);
      console.log(`   Distance: d_entry=${bestParams.d_entry.toFixed(4)}, d_near=${bestParams.d_near.toFixed(4)}`);
      console.log(`   Slope: s_min=${bestParams.s_min.toFixed(6)}`);
      console.log(`   Volatility: rv_cap=${bestParams.rv_cap.toFixed(4)}`);
      console.log(`   Reversion Exits: PT=${bestParams.profitTargetReversion.toFixed(2)}, SL=${bestParams.stopLossReversion.toFixed(2)}, Time=${bestParams.timeStopReversion}s`);
      console.log(`   Trend Exits: PT=${bestParams.profitTargetTrend.toFixed(2)}, SL=${bestParams.stopLossTrend.toFixed(2)}, Time=${bestParams.timeStopTrend}s`);
      
      // Save results
      await this.saveResults(bestParams, bestValue);
      
      return {
        best_parameters: bestParams,
        best_score: bestValue,
        study: this.study,
        all_results: this.results
      };
      
    } catch (error) {
      console.error(`❌ Optimization failed: ${error.message}`);
      throw error;
    }
  }

  /**
   * Save optimization results
   */
  async saveResults(bestParams, bestScore) {
    const fs = require('fs').promises;
    
    const results = {
      optimization_config: this.config,
      best_parameters: bestParams,
      best_score: bestScore,
      all_trials: this.results,
      timestamp: new Date().toISOString(),
      study_stats: {
        n_trials: this.study.trials.length,
        best_trial: this.study.best_trial.number
      }
    };
    
    const filename = `/app/optimization/results/havwap_optimization_${Date.now()}.json`;
    await fs.writeFile(filename, JSON.stringify(results, null, 2));
    
    console.log(`\n💾 Results saved to: ${filename}`);
    
    // Also save best parameters as a new strategy file
    const optimizedStrategy = this.generateOptimizedStrategy(bestParams);
    const strategyFilename = `/app/strategies/havwap-optimized.js`;
    await fs.writeFile(strategyFilename, optimizedStrategy);
    
    console.log(`📄 Optimized strategy saved to: ${strategyFilename}`);
  }

  /**
   * Generate optimized strategy file with best parameters
   */
  generateOptimizedStrategy(params) {
    return `/**
 * HAVWAP Strategy - Optuna Optimized Parameters
 * Generated: ${new Date().toISOString()}
 * Optimization Period: ${this.config.startDate} to ${this.config.endDate}
 * Target Metric: ${this.config.primaryMetric}
 */

const HAVWAPProperStrategy = require('./havwap-proper');

class HAVWAPOptimizedStrategy extends HAVWAPProperStrategy {
  constructor(customParams = {}) {
    // Optuna-optimized parameters
    const optimizedParams = {
      d_entry: ${params.d_entry},
      d_near: ${params.d_near},
      s_min: ${params.s_min},
      rv_cap: ${params.rv_cap},
      iv_chg_cap: ${params.iv_chg_cap},
      theta_imb: ${params.theta_imb},
      deltaTargetReversion: ${params.deltaTargetReversion},
      deltaTargetTrend: ${params.deltaTargetTrend},
      profitTargetReversion: ${params.profitTargetReversion},
      stopLossReversion: ${params.stopLossReversion},
      timeStopReversion: ${params.timeStopReversion},
      profitTargetTrend: ${params.profitTargetTrend},
      stopLossTrend: ${params.stopLossTrend},
      timeStopTrend: ${params.timeStopTrend},
      maxSpreadPct: ${params.maxSpreadPct},
      minVolume: ${params.minVolume},
      zeroDTECloseTime: '${params.zeroDTECloseTime}',
      ...customParams
    };
    
    super(optimizedParams);
    this.name = 'HAVWAP Optimized';
  }
}

module.exports = HAVWAPOptimizedStrategy;`;
  }
}

module.exports = HAVWAPOptunaOptimizer;

File: docker/backtesting-server/optimization/havwap_optuna_optimizer.py
"""
HAVWAP Strategy Optuna Hyperparameter Optimization

Python-based optimization using Optuna to find optimal parameters for HAVWAP strategy.
Interfaces with Node.js backtesting engine via HTTP API.
"""

import optuna
import asyncio
import aiohttp
import json
import numpy as np
import pandas as pd
from datetime import datetime, timedelta
from typing import Dict, List, Tuple, Optional
import logging
import argparse
import os

# Setup logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class HAVWAPOptunaOptimizer:
    def __init__(self, config: Dict):
        self.config = config
        self.session = None
        self.results = []
        
    async def __aenter__(self):
        self.session = aiohttp.ClientSession()
        return self
        
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self.session:
            await self.session.close()

    def define_parameter_spaces(self, trial: optuna.Trial) -> Dict:
        """Define parameter search spaces for Optuna optimization."""
        return {
            # Distance thresholds (most critical parameters)
            'd_entry': trial.suggest_float('d_entry', 0.0005, 0.0050, step=0.0001),
            'd_near': trial.suggest_float('d_near', 0.0002, 0.0015, step=0.0001),
            's_min': trial.suggest_float('s_min', 0.00001, 0.0005, step=0.00001),
            
            # Volatility filters
            'rv_cap': trial.suggest_float('rv_cap', 0.002, 0.008, step=0.0005),
            'iv_chg_cap': trial.suggest_float('iv_chg_cap', 1.5, 5.0, step=0.25),
            
            # Order flow threshold
            'theta_imb': trial.suggest_float('theta_imb', 0.55, 0.80, step=0.05),
            
            # Delta targets
            'deltaTargetReversion': trial.suggest_float('deltaTargetReversion', 0.40, 0.60, step=0.02),
            'deltaTargetTrend': trial.suggest_float('deltaTargetTrend', 0.50, 0.65, step=0.02),
            
            # Exit parameters - REVERSION
            'profitTargetReversion': trial.suggest_float('profitTargetReversion', 0.10, 0.40, step=0.02),
            'stopLossReversion': trial.suggest_float('stopLossReversion', 0.05, 0.20, step=0.01),
            'timeStopReversion': trial.suggest_int('timeStopReversion', 60, 600, step=30),
            
            # Exit parameters - TREND
            'profitTargetTrend': trial.suggest_float('profitTargetTrend', 0.12, 0.45, step=0.02),
            'stopLossTrend': trial.suggest_float('stopLossTrend', 0.08, 0.25, step=0.01),
            'timeStopTrend': trial.suggest_int('timeStopTrend', 120, 900, step=30),
            
            # Option selection
            'maxSpreadPct': trial.suggest_int('maxSpreadPct', 8, 25, step=2),
            'minVolume': trial.suggest_int('minVolume', 0, 5, step=1),
            
            # All-day holding option
            'zeroDTECloseTime': trial.suggest_categorical('zeroDTECloseTime', ['15:30', '15:45', '15:50', '15:55'])
        }

    def get_weekdays(self, start_date: str, end_date: str) -> List[str]:
        """Get list of weekdays between start and end dates."""
        weekdays = []
        current = datetime.strptime(start_date, '%Y-%m-%d')
        end = datetime.strptime(end_date, '%Y-%m-%d')
        
        while current <= end:
            # Monday=0, Sunday=6, so weekdays are 0-4
            if current.weekday() < 5:
                weekdays.append(current.strftime('%Y-%m-%d'))
            current += timedelta(days=1)
            
        return weekdays

    async def run_backtest(self, symbol: str, date: str, params: Dict) -> Optional[Dict]:
        """Run a single backtest via HTTP API."""
        backtest_config = {
            'symbol': symbol,
            'strategy': 'HAVWAP-Proper',
            'startDate': date,
            'endDate': date,
            'contracts': self.config['contracts'],
            'maxPositions': self.config['maxPositions'],
            'initialCapital': self.config['initialCapital'],
            'strategyParams': params
        }
        
        try:
            # Create a modified HAVWAP strategy with trial parameters
            strategy_endpoint = f"{self.config['api_base']}/api/backtest/run"
            
            async with self.session.post(strategy_endpoint, json=backtest_config) as response:
                if response.status == 200:
                    result = await response.json()
                    if result.get('success'):
                        backtest_id = result.get('backtestId')
                        
                        # Poll for completion
                        return await self.wait_for_completion(backtest_id)
                    else:
                        logger.warning(f"Backtest failed for {symbol} {date}: {result.get('message')}")
                        return None
                else:
                    logger.error(f"HTTP error {response.status} for {symbol} {date}")
                    return None
                    
        except Exception as e:
            logger.error(f"Backtest error for {symbol} {date}: {e}")
            return None

    async def wait_for_completion(self, backtest_id: int, timeout: int = 120) -> Optional[Dict]:
        """Wait for backtest completion and return results."""
        start_time = datetime.now()
        
        while (datetime.now() - start_time).seconds < timeout:
            try:
                status_endpoint = f"{self.config['api_base']}/api/backtest/status/{backtest_id}"
                async with self.session.get(status_endpoint) as response:
                    if response.status == 200:
                        status = await response.json()
                        
                        if status.get('status') == 'completed':
                            # Fetch results
                            results_endpoint = f"{self.config['api_base']}/api/backtest/results/{backtest_id}"
                            async with self.session.get(results_endpoint) as results_response:
                                if results_response.status == 200:
                                    return await results_response.json()
                                    
                        elif status.get('status') == 'failed':
                            logger.warning(f"Backtest {backtest_id} failed")
                            return None
                            
                # Wait before polling again
                await asyncio.sleep(2)
                
            except Exception as e:
                logger.error(f"Error polling backtest {backtest_id}: {e}")
                return None
                
        logger.warning(f"Backtest {backtest_id} timed out")
        return None

    def calculate_metrics(self, results: List[Dict]) -> Dict:
        """Calculate performance metrics from backtest results."""
        if not results:
            return {
                'total_return': 0,
                'sharpe_ratio': 0,
                'max_drawdown': 1,
                'profit_factor': 0,
                'win_rate': 0,
                'total_trades': 0,
                'validity_score': 0
            }

        returns = []
        total_pnl = 0
        total_trades = 0
        winning_trades = 0
        gross_wins = 0
        gross_losses = 0
        running_balance = self.config['initialCapital']
        peak = self.config['initialCapital']
        max_drawdown = 0

        for result in results:
            if result and result.get('trades'):
                day_pnl = 0
                
                for trade in result['trades']:
                    trade_pnl = trade.get('realized_pnl', 0)
                    total_pnl += trade_pnl
                    day_pnl += trade_pnl
                    total_trades += 1
                    
                    if trade_pnl > 0:
                        winning_trades += 1
                        gross_wins += trade_pnl
                    else:
                        gross_losses += abs(trade_pnl)
                
                running_balance += day_pnl
                
                # Track drawdown
                if running_balance > peak:
                    peak = running_balance
                else:
                    drawdown = (peak - running_balance) / peak
                    max_drawdown = max(max_drawdown, drawdown)
                
                # Daily return
                daily_return = day_pnl / self.config['initialCapital']
                returns.append(daily_return)
            else:
                returns.append(0)

        # Calculate metrics
        total_return = total_pnl / self.config['initialCapital']
        
        if len(returns) > 1:
            avg_return = np.mean(returns)
            volatility = np.std(returns) * np.sqrt(252)  # Annualized
            sharpe_ratio = avg_return * 252 / volatility if volatility > 0 else 0
        else:
            sharpe_ratio = 0

        win_rate = winning_trades / total_trades if total_trades > 0 else 0
        profit_factor = gross_wins / gross_losses if gross_losses > 0 else (999 if gross_wins > 0 else 0)

        # Validity score
        validity_score = 0
        if total_trades >= self.config['min_trades']:
            validity_score += 0.4
        if max_drawdown <= self.config['max_drawdown']:
            validity_score += 0.3
        if sharpe_ratio > 0:
            validity_score += 0.2
        if profit_factor > 1.0:
            validity_score += 0.1

        return {
            'total_return': total_return,
            'sharpe_ratio': sharpe_ratio,
            'max_drawdown': max_drawdown,
            'profit_factor': profit_factor,
            'win_rate': win_rate,
            'total_trades': total_trades,
            'validity_score': validity_score
        }

    async def objective(self, trial: optuna.Trial) -> float:
        """Objective function for Optuna optimization."""
        logger.info(f"\n🔬 OPTUNA Trial {trial.number}:")
        
        # Get parameter suggestions
        params = self.define_parameter_spaces(trial)
        logger.info(f"   Parameters: d_entry={params['d_entry']:.4f}, d_near={params['d_near']:.4f}, s_min={params['s_min']:.6f}")
        
        # Get test dates
        weekdays = self.get_weekdays(self.config['start_date'], self.config['end_date'])
        logger.info(f"   Testing across {len(weekdays)} trading days")
        
        all_results = []
        total_backtests = 0
        successful_backtests = 0
        
        # Test each symbol across all days
        for symbol in self.config['symbols']:
            for date in weekdays:
                total_backtests += 1
                result = await self.run_backtest(symbol, date, params)
                
                if result:
                    all_results.append(result)
                    successful_backtests += 1
        
        logger.info(f"   ✅ Completed {successful_backtests}/{total_backtests} backtests")
        
        # Calculate metrics
        metrics = self.calculate_metrics(all_results)
        
        logger.info(f"   📊 Results: Return={metrics['total_return']*100:.2f}%, Sharpe={metrics['sharpe_ratio']:.3f}, Trades={metrics['total_trades']}")
        
        # Store result
        self.results.append({
            'trial_number': trial.number,
            'parameters': params,
            'metrics': metrics,
            'successful_backtests': successful_backtests,
            'total_backtests': total_backtests
        })
        
        # Calculate objective score
        if metrics['validity_score'] < 0.5:
            objective_score = -1000
        else:
            if self.config['primary_metric'] == 'sharpe_ratio':
                objective_score = metrics['sharpe_ratio'] * 0.7 + metrics['total_return'] * 0.3
            elif self.config['primary_metric'] == 'total_return':
                objective_score = metrics['total_return'] * 0.7 + metrics['sharpe_ratio'] * 0.3
            else:
                objective_score = np.log(metrics['profit_factor']) * 0.6 + metrics['sharpe_ratio'] * 0.4
            
            # Penalty for excessive drawdown
            if metrics['max_drawdown'] > self.config['max_drawdown']:
                objective_score *= (1 - metrics['max_drawdown'])
        
        logger.info(f"   🎯 Objective Score: {objective_score:.4f}")
        return objective_score

    async def run_optimization(self) -> Dict:
        """Run the Optuna optimization study."""
        logger.info(f"\n🚀 Starting HAVWAP Optuna Optimization:")
        logger.info(f"   Study: {self.config['study_name']}")
        logger.info(f"   Trials: {self.config['n_trials']}")
        logger.info(f"   Period: {self.config['start_date']} to {self.config['end_date']}")
        logger.info(f"   Symbols: {', '.join(self.config['symbols'])}")
        
        # Create study
        study = optuna.create_study(
            study_name=self.config['study_name'],
            direction='maximize',
            sampler=optuna.samplers.TPESampler(
                n_startup_trials=50,
                n_ei_candidates=24,
                seed=42
            ),
            pruner=optuna.pruners.MedianPruner(
                n_startup_trials=25,
                n_warmup_steps=5
            )
        )
        
        # Run optimization
        await asyncio.get_event_loop().run_in_executor(
            None, 
            study.optimize, 
            self.objective, 
            self.config['n_trials']
        )
        
        logger.info(f"\n✅ Optimization completed!")
        
        # Get best parameters
        best_params = study.best_params
        best_value = study.best_value
        
        logger.info(f"\n🏆 BEST PARAMETERS (Score: {best_value:.4f}):")
        logger.info(f"   Distance: d_entry={best_params['d_entry']:.4f}, d_near={best_params['d_near']:.4f}")
        logger.info(f"   Slope: s_min={best_params['s_min']:.6f}")
        logger.info(f"   Volatility: rv_cap={best_params['rv_cap']:.4f}")
        
        # Save results
        self.save_results(best_params, best_value, study)
        
        return {
            'best_parameters': best_params,
            'best_score': best_value,
            'study': study,
            'all_results': self.results
        }

    def save_results(self, best_params: Dict, best_score: float, study: optuna.Study):
        """Save optimization results to file."""
        results = {
            'optimization_config': self.config,
            'best_parameters': best_params,
            'best_score': best_score,
            'all_trials': self.results,
            'timestamp': datetime.now().isoformat(),
            'study_stats': {
                'n_trials': len(study.trials),
                'best_trial': study.best_trial.number
            }
        }
        
        os.makedirs('optimization/results', exist_ok=True)
        filename = f"optimization/results/havwap_optimization_{int(datetime.now().timestamp())}.json"
        
        with open(filename, 'w') as f:
            json.dump(results, f, indent=2)
        
        logger.info(f"💾 Results saved to: {filename}")

async def main():
    """Main optimization function."""
    parser = argparse.ArgumentParser(description='HAVWAP Strategy Optuna Optimization')
    parser.add_argument('--config', help='Configuration file path')
    parser.add_argument('--trials', type=int, default=200, help='Number of optimization trials')
    parser.add_argument('--api-base', default='http://localhost:3002', help='Backtesting API base URL')
    
    args = parser.parse_args()
    
    # Default configuration
    config = {
        'n_trials': args.trials,
        'study_name': f'havwap-opt-{int(datetime.now().timestamp())}',
        'start_date': '2024-10-01',
        'end_date': '2024-10-25',
        'symbols': ['SPY', 'QQQ'],
        'initialCapital': 10000,
        'maxPositions': 1,
        'contracts': 1,
        'primary_metric': 'sharpe_ratio',
        'max_drawdown': 0.15,
        'min_trades': 15,
        'api_base': args.api_base
    }
    
    # Load custom config if provided
    if args.config and os.path.exists(args.config):
        with open(args.config, 'r') as f:
            custom_config = json.load(f)
            config.update(custom_config)
        logger.info(f"✅ Loaded custom config from: {args.config}")
    
    # Run optimization
    async with HAVWAPOptunaOptimizer(config) as optimizer:
        results = await optimizer.run_optimization()
        return results

if __name__ == "__main__":
    asyncio.run(main())